relayer/chains/sui/listener.go

```diff
@@ -2,23 +2,220 @@ package sui
+	var checkpoints []suimodels.CheckpointResponse
+	for _, cp := range checkpointResponses {
+		if len(cp.Transactions) > maxDigestsPerItem {
+			totalBatches := len(cp.Transactions) / maxDigestsPerItem
+			if (len(cp.Transactions) % maxDigestsPerItem) != 0 {
+				totalBatches = totalBatches + 1
+			}
+			for i := 0; i < totalBatches; i++ {
+				fromIndex := i * maxDigestsPerItem
+				toIndex := fromIndex + maxDigestsPerItem
+				if i == totalBatches-1 {
+					toIndex = len(cp.Transactions)
+				}
+				subCheckpoint := suimodels.CheckpointResponse{
+					SequenceNumber: cp.SequenceNumber,
+					Transactions:   cp.Transactions[fromIndex:toIndex],
+				}
+				checkpoints = append(checkpoints, subCheckpoint)
+			}
+		} else {
+			checkpoints = append(checkpoints, cp)
+		}
+	}
+
+	// stage-2: form packets of txDigests
+	var txDigestsList []types.TxDigests
+
+	digests := []string{}
+	fromCheckpoint, _ := strconv.Atoi(checkpoints[0].SequenceNumber)
+	for i, cp := range checkpoints {
+		if (len(digests) + len(cp.Transactions)) > maxDigestsPerItem {
+			toCheckpoint, _ := strconv.Atoi(checkpoints[i-1].SequenceNumber)
+			if len(digests) < maxDigestsPerItem {
+				toCheckpoint, _ = strconv.Atoi(cp.SequenceNumber)
+			}
+			for i, tx := range cp.Transactions {
+				if len(digests) == maxDigestsPerItem {
+					txDigestsList = append(txDigestsList, types.TxDigests{
+						FromCheckpoint: uint64(fromCheckpoint),
+						ToCheckpoint:   uint64(toCheckpoint),
+						Digests:        digests,
+					})
+					digests = cp.Transactions[i:]
+					fromCheckpoint, _ = strconv.Atoi(cp.SequenceNumber)
+					break
+				} else {
+					digests = append(digests, tx)
+				}
+			}
+		} else {
+			digests = append(digests, cp.Transactions...)
 		}
 	}
+
+	lastCheckpointSeq := checkpoints[len(checkpoints)-1].SequenceNumber
+	lastCheckpoint, _ := strconv.Atoi(lastCheckpointSeq)
+	txDigestsList = append(txDigestsList, types.TxDigests{
+		FromCheckpoint: uint64(fromCheckpoint),
+		ToCheckpoint:   uint64(lastCheckpoint),
+		Digests:        digests,
+	})
+
+	return txDigestsList
+}
+
+func (p *Provider) getTxDigestsStream(done chan interface{}, afterSeq string) <-chan types.TxDigests {
+	txDigestsStream := make(chan types.TxDigests)
+
+	go func() {
+		nextCursor := afterSeq
+		checkpointTicker := time.NewTicker(6 * time.Second) //todo need to decide this interval
+
+		for {
+			select {
+			case <-done:
+				return
+			case <-checkpointTicker.C:
+				req := suimodels.SuiGetCheckpointsRequest{
+					Cursor:          nextCursor,
+					Limit:           types.QUERY_MAX_RESULT_LIMIT,
+					DescendingOrder: false,
+				}
+				paginatedRes, err := p.client.GetCheckpoints(context.Background(), req)
+				if err != nil {
+					p.log.Error("failed to fetch checkpoints", zap.Error(err))
+					continue
+				}
+
+				if len(paginatedRes.Data) > 0 {
+					for _, txDigests := range p.GenerateTxDigests(paginatedRes.Data, types.QUERY_MAX_RESULT_LIMIT) {
+						txDigestsStream <- types.TxDigests{
```

Comment:
Does this mean we wait for generating all the tx digests and then only proceed to add them to processing channel ?

Reply:
yes dai, for querying events and to track the checkpoints processed we need to convert paginated checkpoint response to intermediate list of `txDigests{fromCheckpoint, toCheckpoint, digests}`.

Reply:
Can't we process every single checkpoint individually after enough data has been populated for that checkpoint so that processing becomes better and track the checkpoint from that point ? This current implementation might be problematic when we are fetching good number of checkpoint in a single request.